### ISSUE TITLE
Timeout when running migrations from 3.5 to 3.6 leading to corrupted data (migration ran twice)

### ISSUE THREAD

**Issue by tobernguyen** [Kong/kong#12978]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Kong version (`$ kong version`)

Kong 3.6

### Current Behavior

In an attempt to upgrade from Kong 3.5 -> 3.6. It timed out on [this migration](https://github.com/Kong/kong/blob/master/kong/plugins/rate-limiting/migrations/006_350_to_360.lua) for us. I think it's because we have a lot of rate-limiting plugins (~350k rate-limiting rows in `plugins` table).
```
Error: [PostgreSQL error] cluster_mutex callback threw an error: /usr/local/share/lua/5.1/kong/cmd/utils/migrations.lua:104: [PostgreSQL error] failed to run migration '006_350_to_360' up: receive_message: failed to get type: timeout
stack traceback:
	[C]: in function 'assert'
	/usr/local/share/lua/5.1/kong/cmd/utils/migrations.lua:104: in function </usr/local/share/lua/5.1/kong/cmd/utils/migrations.lua:70>
	[C]: in function 'xpcall'
	/usr/local/share/lua/5.1/kong/db/init.lua:403: in function </usr/local/share/lua/5.1/kong/db/init.lua:353>
	[C]: in function 'pcall'
	/usr/local/share/lua/5.1/kong/concurrency.lua:59: in function 'cluster_mutex'
	/usr/local/share/lua/5.1/kong/cmd/utils/migrations.lua:70: in function 'up'
	/usr/local/share/lua/5.1/kong/cmd/migrations.lua:204: in function 'cmd_exec'
	/usr/local/share/lua/5.1/kong/cmd/init.lua:31: in function </usr/local/share/lua/5.1/kong/cmd/init.lua:31>
	[C]: in function 'xpcall'
	/usr/local/share/lua/5.1/kong/cmd/init.lua:31: in function </usr/local/share/lua/5.1/kong/cmd/init.lua:15>
	(command line -e):7: in function 'inline_gen'
	init_worker_by_lua(nginx.conf:232):44: in function <init_worker_by_lua(nginx.conf:232):43>
	[C]: in function 'xpcall'
	init_worker_by_lua(nginx.conf:232):52: in function <init_worker_by_lua(nginx.conf:232):50>

  Run with --v (verbose) or --vv (debug) for more details
migrating core on database 'kong'...
core migrated up to: 022_350_to_360 (executed)
migrating acme on database 'kong'...
acme migrated up to: 003_350_to_360 (executed)
migrating opentelemetry on database 'kong'...
opentelemetry migrated up to: 001_331_to_332 (pending)
migrating rate-limiting on database 'kong'...

--- POD EXITED WITH NON ZERO STATUS ---

--- THEN THE POD IS RESTARTED ---

Error: Database has pending migrations; run 'kong migrations finish'

  Run with --v (verbose) or --vv (debug) for more details
Error: Database has pending migrations; run 'kong migrations finish'
```

I have tried to restore the database to the previous state and try again with `KONG_PG_TIMEOUT=900000 ` (PG Timeout set to 15 mins) but I got the same issue. The error message is the same and it appears only after about 1 mins (definitely less than 15 minutes).

Now, I look into the `plugins` table and I can see that the actual data has been migrated following [this migration state](https://github.com/Kong/kong/blob/master/kong/plugins/rate-limiting/migrations/006_350_to_360.lua) (transforming the flat redis config into a redis object).

Then I look at the `schema_meta` table, this is the values

[query_result_2024-05-04T05_37_58.646724Z.csv](https://github.com/Kong/kong/files/15208193/query_result_2024-05-04T05_37_58.646724Z.csv)

You can see, the `schema_meta` is in a very weird state now:
1. The `rate-limiting` migration is not in a pending state but the `opentelemetry` migration is
2. The `rate-limiting` plugin migration (`006_350_to_360`) ran, and the data was indeed migrated but this migration didn't show up in the `schema_meta` table at all.

At this point, I believe kong data is pretty much messed up but I ran the command suggested just to test (`kong migrations finish`)

```
kong migrations finish
migrating opentelemetry on database 'kong'...
opentelemetry migrated up to: 001_331_to_332 (executed)
1 migration processed
1 executed

New migrations available; run 'kong migrations up' to proceed
```

At this point, I checked the `schema_meta` table again, the `rate-limiting` migrations in `schema_meta` table is still missing `006_350_to_360`. I believe that's why it said "New migrations available".

Now when I follow the instruction to run kong migrations up, it will actually run the rate-limiting migration again (thus running it twice) and resulted in corrupted data

```
kong migrations up
migrating rate-limiting on database 'kong'...
rate-limiting migrated up to: 006_350_to_360 (executed)
migrating response-ratelimiting on database 'kong'...
response-ratelimiting migrated up to: 001_350_to_360 (executed)
2 migrations processed
2 executed
```

I checked the `plugins` table data and can confirm that the data is now completely corrupted (all `rate-limiting` plugins' redis config became `null`).

---
**The migration process seemed to be very confusing because the `schema_meta` can be out of sync with actual migration runs. This seems to be a critical design flaw. I think the migration and the migration metadata update should be wrapped into one transaction so that a migration should never be able to run twice.**

### Expected Behavior

It should finish migration without any errors and should not corrupt the data.

### Steps To Reproduce

1. Upgrade from Kong 3.5 to 3.6 with a lot of rows for `rate-limiting` in `plugins` table (in my case, it is ~350k rate-limiting plugins)

### Anything else?

All the details you can also find in this slack thread: https://kongcommunity.slack.com/archives/CJ6DR0MH7/p1714639347754829

**Comment by hanshuebner** [comment]
Internally tracked in KAG-4416 and KAG-4417

**Comment by hanshuebner** [comment]
This issue will be fixed in the upcoming 3.7 release of Kong Gateway.

**Comment by ms2008** [comment]
@tobernguyen Don't forget to change the `--lock-timeout` arg as well, and it can only be passed in via kong cli args and does not support environment variables.
